Gems/Atom/Asset/Shader/Code/Source/Editor/ShaderAssetBuilder.cpp

```diff
@@ -132,27 +93,22 @@ namespace AZ
         {
             AZStd::string fullPath;
             AzFramework::StringFunc::Path::Join(currentFolderPath.data(), normalizedRelativePath.data(), fullPath);
-            if (!includedFiles.count(fullPath))
-            {
-                includedFiles.insert(fullPath);
-            }
+            includedFiles.insert(fullPath);
             for (const auto &includeDir : includeDirectories)
             {
                 AzFramework::StringFunc::Path::Join(includeDir.c_str(), normalizedRelativePath.data(), fullPath);
-                if (!includedFiles.count(fullPath))
-                {
-                    includedFiles.insert(fullPath);
-                }
+                includedFiles.insert(fullPath);
             }
         }
 
         //! Parses, using depth-first recursive approach, azsl files. Looks for '#include <foo/bar/blah.h>' or '#include "foo/bar/blah.h"' lines
         //! and in turn parses the included files.
         //! The included files are searched in the directories listed in @includeDirectories. Basically it's a similar approach
         //! as how most C-preprocessors would find included files.
-        static void GetListOfIncludedFiles(AZStd::string_view sourceFilePath, const AZStd::vector<AZStd::string>& includeDirectories, AZStd::unordered_set<AZStd::string>& includedFiles)
+        static void GetListOfIncludedFiles(AZStd::string_view sourceFilePath, const AZStd::vector<AZStd::string>& includeDirectories,
+            const ShaderBuilderUtility::IncludedFilesParser& includedFilesParser, AZStd::unordered_set<AZStd::string>& includedFiles)
```

Comment:
Since you introduced a IncludedFilesParser class, why not make the class hold the includedFiles data as a member instead of passing it around?

Reply:
It is less taxing to the reader to understand what the class do. It doesn't need to hold that data and the fact that the functions are "const" makes it really easy to wrap your mind around what does it do.